### PR TITLE
[JENKINS-34996] Release parameters visibility

### PR DESCRIPTION
**Pull request by amuniz**
[JENKINS-34996](https://issues.jenkins-ci.org/browse/JENKINS-34996)

Acknowledges SECURITY-170 (and parent pom upgrade, as it was needed to properly test

@reviewbybees

### CODE REVIEW

**Comment by rsandell** [comment]
It looks like a bit of strange mix of solving the issue.

You could either skip adding the parameters and provide them via an EnvironmentContributor instead.

Or if, for legacy reasons with external scripts for example, you need the parameters still there you should use the new constructor added in 2.6 `ParameterAction(parameters, additionalSafeParameters)` - You could call that with reflection to not have to depend on a newer core.

**Comment by rsandell** [comment]
Using a new ParametersAction will cause the xml api to be scewed anyway because it is not the old params in the xml api iirc.

**Comment by amuniz** [comment]
> You could either skip adding the parameters and provide them via an EnvironmentContributor instead.

This causes regressions is the build page (missing parameters).

> If [...] you need the parameters still there you should use the new constructor added in 2.6

Yeah, I considered this option but, if possible, I prefer to avoid the reflection dance (and the lapse of  core versions where it won't work).

> Using a new ParametersAction will cause the xml api to be scewed anyway because it is not the old params in the xml api iirc.

No, there are no XML API incompatibilities. The change is compatible as this `ParametersAction` override generates:

```xml
<action>
    <parameter>
        <name>MY_PARAM</name>
        <value>4567</value>
    </parameter>
</action>
```

Which is exactly the same than before (`@Exported`/`@ExportedBean` are inherited).

File: pom.xml [hunk 1]
```diff
@@ -13,17 +13,12 @@
...
-        <artifactId>maven-release-plugin</artifactId>
-        <version>2.5.1</version>
-        <configuration>
-          <goals>deploy</goals>
-        </configuration>
-      </plugin>
-    </plugins>
-  </build>
+  <properties>
+    <jenkins.version>1.609.3</jenkins.version>
```
Comment (oleg-nenashev) [inline pom.xml]: So it bumps the version. What's the reason? New API dependency?

File: pom.xml [hunk 2]
```diff
@@ -43,6 +38,7 @@
   	<dependency>
   		<groupId>org.jenkins-ci.main</groupId>
   		<artifactId>maven-plugin</artifactId>
+        <version>2.13</version>
```
Comment (oleg-nenashev) [inline pom.xml]: Spaces vs. tabs

File: src/main/java/hudson/plugins/release/SafeParametersAction.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+import hudson.EnvVars;
+import hudson.Extension;
+import hudson.model.EnvironmentContributor;
+import hudson.model.ParameterValue;
+import hudson.model.ParametersAction;
+import hudson.model.Run;
+import hudson.model.TaskListener;
+
+@Restricted(NoExternalUse.class)
+public class SafeParametersAction extends ParametersAction {
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/release/SafeParametersAction.java]: I would not do it public at all. Restricted does not protect from misusages in Pipelines and system groovy scripts

Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/release/SafeParametersAction.java]: NIT in any case

File: src/main/java/hudson/plugins/release/SafeParametersAction.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+    private final List<ParameterValue> parameters;
+
+    /**
+     * At this point the list of parameter values is granted to be safe, which is
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/release/SafeParametersAction.java]: granted/guaranteed ?

File: pom.xml (hunk 1 above)
Comment (amuniz) [inline pom.xml]: No, it's common sense. 1.481 was not even Jenkins, and 1.609.3 is the oldest LTS in a 1 year window. So I think this bump is not going to affect anyone.

File: pom.xml (hunk 2 above)
Comment (amuniz) [inline pom.xml]: They are already mixed in this plugin, I just used spaces.

File: src/main/java/hudson/plugins/release/SafeParametersAction.java (hunk 4 above)
Comment (amuniz) [inline src/main/java/hudson/plugins/release/SafeParametersAction.java]: Yeah, guaranteed, will change it.

**Comment by oleg-nenashev** [comment]
I'm aware that <code>SafeParametersAction</code> does not actually override all methods using <code>parameters</code> in 1.609.3. E.g. createBuildWrappers, createVariableResolver, etc. I am not sure such change covers all use-cases. 

Maybe it applies to my fix in https://github.com/jenkinsci/promoted-builds-plugin/pull/93

**Comment by amuniz** [comment]
> I'm aware that SafeParametersAction does not actually override all methods using parameters in 1.609.3. E.g. createBuildWrappers, createVariableResolver.

I don't see uses of that methods in the scope of this plugin, and any external call must follow the rules of `ParametersAction` - except for `getParameters` and `getParameter(String)`.

Moreover, I've added test (if you think there is some uncovered case, I'm ok to add it) and I've tested it manually (including an upgrade).

So, if you still feel fear, what could I do? :smile:

File: pom.xml (hunk 1 above)
Comment (oleg-nenashev) [inline pom.xml]: It's debatable. My previous instance still work on 1.509.4, @olivergondza uses 1.580.x IIRC.
For enterprise installations it's common to be behind the trends due to [stabilization/migration testing] efforts. 

So I would ask the plugin maintainer

File: src/main/resources/hudson/plugins/release/promotion/ReleasePromotionCondition/index.jelly [hunk 5]
```diff
@@ -1,3 +1,4 @@
+<?jelly escape-by-default='true'?>
```
Comment (oleg-nenashev) [inline src/main/resources/hudson/plugins/release/promotion/ReleasePromotionCondition/index.jelly]: NIT: This file does not need to be Jelly at all. Not related to this PR

**Comment by oleg-nenashev** [comment]
> So, if you still feel fear, what could I do?

At least the fix does not decrease the technical debt (it fixes some cases at least), so I'm not going to force any major rework in this PR. My suggestion would be to migrate Jenkins core code to getParameters() getter, for example. CC @rsandell

File: src/main/resources/hudson/plugins/release/promotion/ReleasePromotionCondition/index.jelly (hunk 5 above)
Comment (amuniz) [inline src/main/resources/hudson/plugins/release/promotion/ReleasePromotionCondition/index.jelly]: Mmm, I think `InjectedTest` was complaining about the lack of this line.

File: pom.xml (hunk 1 above)
Comment (amuniz) [inline pom.xml]: Ok.

@avalanche123 not sure if you are the current maintainer but you seem to have done the last releases. Are you acting as maintainer?

**Comment by amuniz** [comment]
@oleg-nenashev @rsandell ping?

**Comment by sevilS** [comment]
I think that @Fiouz, @avalanche123  are the most recent maintainers of release-plugin repo. ¿There is any plan about merge this pull request?

**Comment by recampbell** [comment]
Perhaps @oleg-nenashev and @rsandell can review now that they're back from vacations.

**Comment by raspy** [comment]
When can we expect plugin release with this fix? This has been impacting our release procedures for quite some time now.

**Comment by nikkobertrand** [comment]
Our release procedure relied on this and is quite broken now.

**Comment by zeraturl** [comment]
Hi, can you estimate when you merge the fix and release a new plugin version?

Or can you descripe how to deploy the fix of the plugin to a running jenkins server?
Thx.

**Comment by noppelmax** [comment]
With this pull request the release plugin is working again for me. @zeraturl you may run `git pull origin pull/17/head` and `mvn package`. This gives you the .hpi which can be loaded into jenkins. Works for me. Will jump back to official builds if this one is in.

**Comment by oleg-nenashev** [comment]
Well, @avalanche123 does not seem to be active on GitHub. I'll go forward and release this fix

**Pull request merged**